File: collect_app/src/androidTestBranded/java/io/ffem/collect/android/AutoFormTest.java
package io.ffem.collect.android;

import android.Manifest;
import android.app.Instrumentation.ActivityResult;
import android.content.Context;
import android.content.Intent;
import android.os.SystemClock;
import android.view.View;

import androidx.test.espresso.contrib.RecyclerViewActions;
import androidx.test.rule.GrantPermissionRule;

import net.bytebuddy.utility.RandomString;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;
import org.junit.BeforeClass;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.odk.collect.android.R;
import org.odk.collect.android.injection.config.AppDependencyModule;
import org.odk.collect.android.support.CopyFormRule;
import org.odk.collect.android.support.FormActivityTestRule;
import org.odk.collect.android.support.ResetStateRule;
import org.odk.collect.android.utilities.ActivityAvailability;

import java.util.ArrayList;

import tools.fastlane.screengrab.Screengrab;
import tools.fastlane.screengrab.UiAutomatorScreenshotStrategy;
import tools.fastlane.screengrab.locale.LocaleTestRule;

import static android.app.Activity.RESULT_OK;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.swipeLeft;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.intent.Intents.intending;
import static androidx.test.espresso.intent.matcher.IntentMatchers.hasAction;
import static androidx.test.espresso.intent.matcher.IntentMatchers.hasExtra;
import static androidx.test.espresso.matcher.ViewMatchers.hasSibling;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.Matchers.startsWith;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Integration test that runs through a form with all question types.
 * <p>
 * <a href="https://docs.fastlane.tools/actions/screengrab/"> screengrab </a> is used to generate screenshots for
 * documentation and releases. Calls to Screengrab.screenshot("image-name") trigger screenshot
 * creation.
 */
public class AutoFormTest {

    @ClassRule
    public static final LocaleTestRule LOCALE_TEST_RULE = new LocaleTestRule();
    private static final String AUTO_TEST_FORM = "auto-test.xml";
    public final AutoFormTestRule activityTestRule = new AutoFormTestRule();
    private final ActivityAvailability activityAvailability = mock(ActivityAvailability.class);
    @Rule
    public RuleChain copyFormChain = RuleChain
            .outerRule(GrantPermissionRule.grant(
                    Manifest.permission.READ_EXTERNAL_STORAGE,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE)
            )
            .around(new ResetStateRule(new AppDependencyModule() {
                @Override
                public ActivityAvailability providesActivityAvailability(Context context) {
                    return activityAvailability;
                }
            }))
            .around(new CopyFormRule(AUTO_TEST_FORM))
            .around(activityTestRule);

    @BeforeClass
    public static void beforeAll() {
        Screengrab.setDefaultScreenshotStrategy(new UiAutomatorScreenshotStrategy());
    }

    public static Matcher<View> withIndex(final Matcher<View> matcher, final int index) {
        return new TypeSafeMatcher<View>() {
            int currentIndex = 0;

            @Override
            public void describeTo(Description description) {
                description.appendText("with index: ");
                description.appendValue(index);
                matcher.describeTo(description);
            }

            @Override
            public boolean matchesSafely(View view) {
                return matcher.matches(view) && currentIndex++ == index;
            }
        };
    }

    @Test
    public void testActivityOpen() {

        int textIndex = 0;

        ArrayList<String> extras = new ArrayList<>();

        onView(withText(R.string.form_forward)).perform(click());

        extras.add("value");
        testExternal("io.ffem.water", "Arsenic (0 - 500)", extras,
                textIndex++, 0, false);
        extras.clear();

        extras.add("value");
        testExternal("io.ffem.water", "Arsenic (0 - 4000)", extras,
                textIndex, 1, true);
        extras.clear();

        test_pH();

        testBoron();

        extras.add("value");
        testExternal("io.ffem.soil", "Calcium and Magnesium", extras,
                4, 0, true);
        extras.clear();

        onView(withText(R.string.form_forward)).perform(click());

        extras.add("value");
        testExternal("io.ffem.water", "Fluoride", extras,
                2, 2, true);
        extras.clear();

        testFreeChlorine();

        testTotalIron();

        testEnd();
    }

    private void testBoron() {
        onView(withText(startsWith("Available Boron"))).perform(swipeLeft());
    }

    private void test_pH() {
        onView(withText(startsWith("pH"))).perform(swipeLeft());
    }

    private void testFreeChlorine() {
        onView(withText(startsWith("Free Chlorine"))).perform(swipeLeft());
    }

    private void testExternal(String action, String title, ArrayList<String> extras,
                              int testIndex, int buttonIndex, boolean goNext) {

        when(activityAvailability.isActivityAvailable(any(Intent.class)))
                .thenReturn(false);

        onView(withIndex(withText(R.string.launch_app), buttonIndex)).perform(click());

        onView(withId(android.R.id.button2)).perform(click());

        when(activityAvailability.isActivityAvailable(any(Intent.class)))
                .thenReturn(true);

        openWidgetList();
        onView(withId(R.id.list)).perform(RecyclerViewActions.actionOnItemAtPosition(testIndex, click()));

        // Replace with Intent value:
        String exStringWidgetSecondText = randomString();

        Intent stringIntent = new Intent();
        for (String extra : extras) {
            stringIntent.putExtra(extra + "", exStringWidgetSecondText);
        }

        ActivityResult exStringResult = new ActivityResult(RESULT_OK, stringIntent);

        switch (extras.size()) {
            case 1:
                intending(allOf(
                        hasAction(action),
                        hasExtra(extras.get(0) + "", null)
                )).respondWith(exStringResult);
                break;
            case 2:
                intending(allOf(
                        hasAction(action),
                        hasExtra(extras.get(0) + "", null),
                        hasExtra(extras.get(1) + "", null)
                )).respondWith(exStringResult);
                break;
            case 3:
                intending(allOf(
                        hasAction(action),
                        hasExtra(extras.get(0) + "", null),
                        hasExtra(extras.get(1) + "", null),
                        hasExtra(extras.get(2) + "", null)
                )).respondWith(exStringResult);
                break;
            case 4:
                intending(allOf(
                        hasAction(action),
                        hasExtra(extras.get(0) + "", null),
                        hasExtra(extras.get(1) + "", null),
                        hasExtra(extras.get(2) + "", null),
                        hasExtra(extras.get(3) + "", null)
                )).respondWith(exStringResult);
                break;
        }

        when(activityAvailability.isActivityAvailable(any(Intent.class)))
                .thenReturn(true);

        if (extras.size() == 1 && extras.get(0).equals("value")) {
            onView(withIndex(withId(R.id.simple_button), buttonIndex)).perform(click());
            onView(withText(exStringWidgetSecondText)).check(matches(isDisplayed()));
        } else {
            onView(allOf(withText(R.string.launch_app), hasSibling(withText(title)))).perform(click());
            onView(withIndex(withText(extras.get(0)
                    .replace("_", " ") + ": "), 0)).check(matches(isDisplayed()));
            onView(withIndex(withText(exStringWidgetSecondText), 0)).check(matches(isDisplayed()));
        }

        openWidgetList();
        onView(withId(R.id.list)).perform(RecyclerViewActions.actionOnItemAtPosition(testIndex, click()));

        if (extras.size() == 1 && extras.get(0).equals("value")) {
            onView(withText(exStringWidgetSecondText)).check(matches(isDisplayed()));
        } else {
            onView(withIndex(withText(exStringWidgetSecondText), 0)).check(matches(isDisplayed()));
            onView(withIndex(withText(extras.get(0)
                    .replace("_", " ") + ": "), 0)).check(matches(isDisplayed()));
        }

        if (goNext) {
            onView(withIndex(withText(startsWith(title)), 0)).perform(swipeLeft());
        }
    }

    private void testTotalIron() {
        onView(withText(startsWith("Total Iron"))).perform(swipeLeft());
    }

    private void openWidgetList() {
        SystemClock.sleep(1000);
        onView(withId(R.id.menu_goto)).perform(click());
        SystemClock.sleep(1000);
    }

    private String randomString() {
        return RandomString.make();
    }

    private void testEnd() {
        onView(withText(activityTestRule.getActivity().getString(R.string.save_enter_data_description,
                "Automated Testing"))).check(matches(isDisplayed()));
        onView(withText(R.string.save_form_as)).check(matches(isDisplayed()));
        onView(withText(R.string.mark_finished)).perform(click());
        onView(withText(R.string.quit_entry)).check(matches(isDisplayed()));
    }

    private static class AutoFormTestRule extends FormActivityTestRule {
        AutoFormTestRule() {
            super(AUTO_TEST_FORM);
        }
    }
}

File: collect_app/src/androidTestBranded/java/io/ffem/collect/android/AutoFormTest.kt
package io.ffem.collect.android

import android.app.Activity
import android.app.Instrumentation
import android.content.Context
import android.content.Intent
import android.os.SystemClock
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.contrib.RecyclerViewActions
import androidx.test.espresso.intent.Intents
import androidx.test.espresso.intent.matcher.IntentMatchers
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.GrantPermissionRule
import androidx.test.uiautomator.UiDevice
import io.ffem.collect.android.util.TestHelper
import io.ffem.collect.android.util.mDevice
import net.bytebuddy.utility.RandomString
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.Matchers
import org.hamcrest.TypeSafeMatcher
import org.junit.BeforeClass
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.odk.collect.android.R
import org.odk.collect.android.injection.config.AppDependencyModule
import org.odk.collect.android.support.CopyFormRule
import org.odk.collect.android.support.FormActivityTestRule
import org.odk.collect.android.support.ResetStateRule
import org.odk.collect.android.utilities.ActivityAvailability
import java.util.*

class AutoFormTest {
    val activityTestRule = AutoFormTestRule()
    private val activityAvailability = Mockito.mock(ActivityAvailability::class.java)

    @Rule
    @JvmField
    var copyFormChain: RuleChain = RuleChain
            .outerRule(GrantPermissionRule.grant()
            )
            .around(ResetStateRule(object : AppDependencyModule() {
                override fun providesActivityAvailability(context: Context): ActivityAvailability {
                    return activityAvailability
                }
            }))
            .around(CopyFormRule(AUTO_TEST_FORM))
            .around(activityTestRule)

    @Test
    fun testActivityOpen() {
        val extras = ArrayList<String>()
        onView(ViewMatchers.withText(R.string.form_forward)).perform(ViewActions.click())
        onView(ViewMatchers.withText(R.string.form_forward)).perform(ViewActions.click())
        test_pH()
        testBoron()
        extras.add("value")
        testExternal("io.ffem.soil", "Calcium & Magnesium", extras,
                4, 0, true)
        extras.clear()
        onView(ViewMatchers.withText(R.string.form_forward)).perform(ViewActions.click())
        extras.add("value")
        testExternal("io.ffem.water", "Fluoride", extras,
                2, 2, true)
        extras.clear()
        testFreeChlorine()
        testTotalIron()
        testEnd()
    }

    private fun testBoron() {
        onView(ViewMatchers.withText(Matchers.startsWith("Available Boron"))).perform(ViewActions.swipeLeft())
    }

    private fun test_pH() {
        onView(ViewMatchers.withText(Matchers.startsWith("pH"))).perform(ViewActions.swipeLeft())
    }

    private fun testFreeChlorine() {
        onView(ViewMatchers.withText(Matchers.startsWith("Free Chlorine"))).perform(ViewActions.swipeLeft())
    }

    @Suppress("SameParameterValue")
    private fun testExternal(action: String, title: String, extras: ArrayList<String>,
                             testIndex: Int, buttonIndex: Int, goNext: Boolean) {
        Mockito.`when`(activityAvailability.isActivityAvailable(ArgumentMatchers.any(Intent::class.java)))
                .thenReturn(false)
        onView(withIndex(ViewMatchers.withText(title), buttonIndex)).perform(ViewActions.click())
        onView(ViewMatchers.withId(android.R.id.button2)).perform(ViewActions.click())
        Mockito.`when`(activityAvailability.isActivityAvailable(ArgumentMatchers.any(Intent::class.java)))
                .thenReturn(true)
        openWidgetList()
        onView(ViewMatchers.withId(R.id.list)).perform(RecyclerViewActions.actionOnItemAtPosition<RecyclerView.ViewHolder>(testIndex, ViewActions.click()))

        // Replace with Intent value:
        val exStringWidgetSecondText = randomString()
        val stringIntent = Intent()
        for (extra in extras) {
            stringIntent.putExtra(extra + "", exStringWidgetSecondText)
        }
        val exStringResult = Instrumentation.ActivityResult(Activity.RESULT_OK, stringIntent)
        when (extras.size) {
            1 -> Intents.intending(Matchers.allOf(
                    IntentMatchers.hasAction(action),
                    IntentMatchers.hasExtra<Any?>(extras[0] + "", null)
            )).respondWith(exStringResult)
            2 -> Intents.intending(Matchers.allOf(
                    IntentMatchers.hasAction(action),
                    IntentMatchers.hasExtra<Any?>(extras[0] + "", null),
                    IntentMatchers.hasExtra<Any?>(extras[1] + "", null)
            )).respondWith(exStringResult)
            3 -> Intents.intending(Matchers.allOf(
                    IntentMatchers.hasAction(action),
                    IntentMatchers.hasExtra<Any?>(extras[0] + "", null),
                    IntentMatchers.hasExtra<Any?>(extras[1] + "", null),
                    IntentMatchers.hasExtra<Any?>(extras[2] + "", null)
            )).respondWith(exStringResult)
            4 -> Intents.intending(Matchers.allOf(
                    IntentMatchers.hasAction(action),
                    IntentMatchers.hasExtra<Any?>(extras[0] + "", null),
                    IntentMatchers.hasExtra<Any?>(extras[1] + "", null),
                    IntentMatchers.hasExtra<Any?>(extras[2] + "", null),
                    IntentMatchers.hasExtra<Any?>(extras[3] + "", null)
            )).respondWith(exStringResult)
        }
        Mockito.`when`(activityAvailability.isActivityAvailable(ArgumentMatchers.any(Intent::class.java)))
                .thenReturn(true)
        if (extras.size == 1 && extras[0] == "value") {
            onView(withIndex(ViewMatchers.withId(R.id.simple_button), buttonIndex)).perform(ViewActions.click())
            onView(ViewMatchers.withText(exStringWidgetSecondText)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        } else {
            onView(Matchers.allOf(ViewMatchers.withText(title), ViewMatchers.hasSibling(ViewMatchers.withText(title)))).perform(ViewActions.click())
            onView(withIndex(ViewMatchers.withText(extras[0]
                    .replace("_", " ") + ": "), 0)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
            onView(withIndex(ViewMatchers.withText(exStringWidgetSecondText), 0)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        }
        openWidgetList()
        onView(ViewMatchers.withId(R.id.list)).perform(RecyclerViewActions.actionOnItemAtPosition<RecyclerView.ViewHolder>(testIndex, ViewActions.click()))
        if (extras.size == 1 && extras[0] == "value") {
            onView(ViewMatchers.withText(exStringWidgetSecondText)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        } else {
            onView(withIndex(ViewMatchers.withText(exStringWidgetSecondText), 0)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
            onView(withIndex(ViewMatchers.withText(extras[0]
                    .replace("_", " ") + ": "), 0)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        }
        if (goNext) {
            onView(withIndex(ViewMatchers.withText(Matchers.startsWith(title)), 0)).perform(ViewActions.swipeLeft())
        }
    }

    private fun testTotalIron() {
        onView(ViewMatchers.withText(Matchers.startsWith("Total Iron"))).perform(ViewActions.swipeLeft())
    }

    private fun openWidgetList() {
        SystemClock.sleep(1000)
        onView(ViewMatchers.withId(R.id.menu_goto)).perform(ViewActions.click())
        SystemClock.sleep(1000)
    }

    private fun randomString(): String {
        return RandomString.make()
    }

    private fun testEnd() {
        onView(ViewMatchers.withText(activityTestRule.activity.getString(R.string.save_enter_data_description,
                "Automated Testing"))).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        onView(ViewMatchers.withText(R.string.save_form_as)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        onView(ViewMatchers.withText(R.string.mark_finished)).perform(ViewActions.click())
        onView(ViewMatchers.withText(R.string.quit_entry)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    class AutoFormTestRule internal constructor() : FormActivityTestRule(AUTO_TEST_FORM)
    companion object {

        private const val AUTO_TEST_FORM = "auto-test.xml"

        fun withIndex(matcher: Matcher<View?>, index: Int): Matcher<View> {
            return object : TypeSafeMatcher<View>() {
                var currentIndex = 0
                override fun describeTo(description: Description) {
                    description.appendText("with index: ")
                    description.appendValue(index)
                    matcher.describeTo(description)
                }

                public override fun matchesSafely(view: View?): Boolean {
                    return matcher.matches(view) && currentIndex++ == index
                }
            }
        }

        @JvmStatic
        @BeforeClass
        fun initialize() {
            if (!TestHelper.isDeviceInitialized()) {
                mDevice = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
            }
        }
    }
}